### PR TITLE
fix(skins): validate the token bridge against all local interfaces

### PR DESCRIPTION
**Pull request by allofmeng**
## Summary

- Problem: the skin token bridge trusted only `localhost` or one WiFi IP cached at startup.
- Why it matters: legitimate Ethernet, secondary-adapter, and renewed DHCP addresses could lose account-proxy token injection.
- What changed: validate against the device's current interface addresses, with an instance-injected provider; use the cached WiFi IP only if enumeration fails.
- What did NOT change: loopback support, URI structure checks, arbitrary-host rejection, REST/WebSocket contracts, and BLE/USB behavior.

## Change Type (select all)

- [x] Bug fix
- [ ] Feature
- [x] Refactor required for the fix
- [x] Docs
- [x] Security hardening
- [ ] Chore / infra
- [ ] Plugin (DYE2 or bundled skin)

## Scope (select all touched areas)

- [ ] BLE transport / device comms
- [ ] REST API / handlers
- [ ] WebSocket API
- [ ] Machine state / shot logic
- [ ] Scale / weight / flow
- [ ] Profiles / beans / grinders / workflows
- [x] WebUI skins
- [ ] Plugins / JS runtime
- [ ] UI / Flutter widgets
- [ ] Storage / Drift database
- [ ] CI / build / infra
- [x] Docs / specs

## Linked Issues

- Closes: N/A
- Related: #300, #337

## Root Cause (if bug fix)

- Root cause: token injection compared the request host with one WiFi address cached for display at server startup.
- Missing detection or guardrail: tests did not cover secondary interfaces, stale cached addresses, or interface-enumeration failure.
- Contributing context: the display address and security allowlist had shared the same cached value despite different freshness requirements.

## Regression Test Plan (if bug fix or refactor)

- Coverage level that should have caught this:
  - [x] Unit test
  - [ ] Integration test (mock transport edge)
  - [ ] End-to-end test (`simulate=1` + curl/websocat)
  - [ ] Existing coverage already sufficient
- Target test or file: `test/webui_support/webui_token_injection_test.dart`
- Scenario the test should lock in: accept a current secondary-interface IP; reject a stale cached WiFi IP after successful enumeration; accept the cached IP only when enumeration throws; continue rejecting arbitrary hosts.
- If no new test added, why not: N/A

## Documentation Obligations (required)

- [ ] API spec updated: `assets/api/rest_v1.yml` or `assets/api/websocket_v1.yml` (no REST/WebSocket contract change)
- [ ] API docs updated: `doc/Api.md` (no endpoint change)
- [ ] Plugin docs updated: `doc/Plugins.md` (no plugin API change)
- [x] Skin docs updated: `doc/Skins.md` (token-injection host behavior documented)
- [ ] Profile docs updated: `doc/Profiles.md` (no profile change)
- [ ] Device docs updated: `doc/DeviceManagement.md` (no device-flow change)
- [ ] N/A — no docs affected

Additional internal rationale is documented in `doc/AI_API_NOTES.md`.

## Security Impact (required)

- New or changed REST endpoints? (`No`)
- New or changed WebSocket topics? (`No`)
- New or changed network calls? (`Yes`)
- BLE/USB surface changed? (`No`)
- File system access changed? (`No`)
- Plugin sandbox boundary changed? (`No`)
- If any `Yes`, explain risk and mitigation: port 3000 token injection now checks all currently assigned local IPs. Successful enumeration is authoritative, so stale cached DHCP addresses are rejected; loopback remains accepted; arbitrary hostnames remain rejected to preserve the DNS-rebinding boundary. The cached WiFi IP is used only when local interface enumeration fails.

## User-Visible Changes

Skins reached through a currently assigned Ethernet, secondary-adapter, or renewed DHCP address receive the account-proxy token as expected.

### Local gates (run before pushing)

- [x] `flutter analyze` — clean (no new warnings)
- [x] `flutter test` — all 2458 tests pass
- [ ] `(cd packages/dye2-plugin && npm run build)` — N/A; no plugin changes

### Manual verification (if applicable)

- OS / platform tested: macOS test host
- Simulated devices? (`simulate=1`): `No`
- Real hardware? (DE1/Bengle/scale): `No`
- What you personally verified and how: local HTTP server tests sent explicit Host headers for current, stale, fallback, loopback, and arbitrary-host cases.
- Edge cases checked: secondary interface, DHCP-stale cache, enumeration failure, untrusted hostname.
- What you did **not** verify: real multi-adapter hardware or mobile OS interface enumeration.

### Evidence

- [x] Test output (failing before + passing after)
- [ ] Log snippets
- [ ] Screenshot / recording (UI changes)
- [ ] curl / websocat output (API changes)

Targeted suite: 16 tests passed. Full suite: 2458 tests passed. `flutter analyze`: no issues found.

## Compatibility & Migration

- Backward compatible? (`Yes`)
- Config / env changes needed? (`No`)
- Database migration needed? (`No`)
- If any `No` or `Yes`, explain exact steps: no migration or configuration changes are required.

## Risks & Mitigations

- Risk: interface enumeration can fail on a platform.
  - Mitigation: fall back to the cached WiFi address only for that failure path; loopback remains available.

### CODE REVIEW

File: lib/src/webui_support/webui_service.dart [hunk 1]
```diff
@@ -142,15 +146,48 @@ class WebUIService {
...
+  /// next process restart. NetworkInterface.list() is a local OS call (no
+  /// network I/O) and this path only runs on HTML responses, so a live lookup
+  /// per request is cheap enough not to need a cache.
+  Future<bool> _isLocalHost(String host) async {
+    if (host == 'localhost' || host == '127.0.0.1' || host == '::1') {
+      return true;
+    }
+    if (host == _localIP) return true;
+    try {
+      final addresses = await listLocalAddresses();
```
Comment (tadelv) [inline lib/src/webui_support/webui_service.dart]: `_localIP` is cached once with `??=` and may no longer be assigned after DHCP renewal or a network switch. Accepting it before live enumeration means this security allowlist retains stale addresses, despite the helper claiming to validate current bindings.

Please make the live interface list authoritative when enumeration succeeds, and use `_localIP` only as a fallback when enumeration throws. Add regression tests proving that a stale cached IP is rejected and that the cached IP is accepted only on enumeration failure.

File: lib/src/webui_support/webui_service.dart [hunk 2]
```diff
@@ -118,8 +112,18 @@ class WebUIService {
...
-  static Future<String?> Function() resolveWifiIP =
-      NetworkInfo().getWifiIP;
+  static Future<String?> Function() resolveWifiIP = NetworkInfo().getWifiIP;
+
+  /// Test seam: override in tests to control which addresses [_isLocalHost]
+  /// accepts, without depending on the test runner's real network interfaces.
+  @visibleForTesting
+  static Future<List<String>> Function() listLocalAddresses = () async {
+    final interfaces = await NetworkInterface.list(includeLoopback: false);
+    return [
```
Comment (tadelv) [inline lib/src/webui_support/webui_service.dart]: Please avoid adding another mutable static callback. This is process-global state governing a security boundary, and tests have to restore it by duplicating the production implementation.

Inject the address provider into `WebUIService` as a final instance dependency with a production default. That follows the repository's constructor-injection rule, isolates tests, and removes reset/drift hazards.

File: lib/src/webui_support/webui_service.dart [hunk 3]
```diff
@@ -142,15 +146,48 @@ class WebUIService {
...
+  /// Deliberately NOT interface-cached: WebUIService.serveFolderAtPath caches
+  /// [_localIP] once (a single WiFi IP, for the QR/hero-card display, which
+  /// tolerates staleness) but the token bridge must not — a DHCP renewal, a
+  /// network switch, or simply reaching this device over Ethernet instead of
+  /// WiFi must not silently strand every skin's account-proxy calls until the
+  /// next process restart. NetworkInterface.list() is a local OS call (no
+  /// network I/O) and this path only runs on HTML responses, so a live lookup
+  /// per request is cheap enough not to need a cache.
+  Future<bool> _isLocalHost(String host) async {
+    if (host == 'localhost' || host == '127.0.0.1' || host == '::1') {
```
Comment (tadelv) [inline lib/src/webui_support/webui_service.dart]: This is too much incident history, threat-model explanation, and performance rationale in production code, and it conflicts with the explicit `AGENTS.md` guidance against explanatory comments in new code.

Please reduce this to a one-line contract comment and move the rationale to the relevant `doc/AI_*` note (the Auth Proxy section in `doc/AI_API_NOTES.md` is a natural fit).

File: lib/src/webui_support/webui_service.dart [hunk 4]
```diff
@@ -142,15 +146,48 @@ class WebUIService {
...
     if (uri.scheme != 'http' ||
         uri.port != port ||
         uri.userInfo.isNotEmpty ||
-        (uri.host != 'localhost' && uri.host != _localIP)) {
+        !await _isLocalHost(uri.host)) {
+      _log.info(
+        'Skin API bridge not injected for untrusted host "${uri.host}"',
+      );
       return null;
     }
```
Comment (tadelv) [inline lib/src/webui_support/webui_service.dart:179]: The log message says the host is untrusted even when rejection was caused by the scheme, port, or user-info checks. It will also emit an info-level entry for every rejected HTML request.

Please separate structural URI validation from host validation, and log only a genuine host rejection—preferably at a lower diagnostic level, or omit the per-request message.

**Review by tadelv — CHANGES_REQUESTED** [review]
The direction is sensible and CI is green, but I do not think this is ready to merge yet.

Required before merge:

1. Make the current interface list authoritative when enumeration succeeds. `_localIP` is cached with `??=` and may be stale after DHCP renewal or a network switch, so accepting it unconditionally preserves an obsolete entry in a security allowlist. Keep `_localIP` only as a fallback when enumeration fails, and add tests for both the stale-address rejection and enumeration-failure fallback.
2. Do not add a mutable process-global test seam for a security check. Inject the address provider into `WebUIService` as an instance dependency with a production default.
3. Complete the repository hard gates: use the PR template, fill in the required security/documentation sections, and update `doc/Skins.md`. This changes network-facing token-injection behavior, and the repository explicitly requires skin-behavior documentation in the same PR.
4. Move the threat-model/history explanation out of production code and into the appropriate `doc/AI_*` documentation. `AGENTS.md` explicitly asks new code to use clear names and small functions rather than explanatory comments.
5. Revert unrelated formatter churn so the diff stays surgical.

The implementation should remain small: one instance-owned address provider, a short local-host predicate, two additional regression cases, and concise documentation.

**Review by tadelv — APPROVED** [review]
Re-reviewed after the requested changes.

The live interface list is now authoritative when enumeration succeeds, the cached WiFi IP is fallback-only, the address provider is instance-injected, regression coverage includes stale-address rejection and enumeration-failure fallback, and the documentation and diff-scope requirements are addressed. URI validation and host validation are also cleanly separated.

CI is green. No remaining blocking findings.
